### PR TITLE
component in prefab node add __prefab attr

### PR DESCRIPTION
**Pull request by arsen2010**
Re: cocos-creator/3d-tasks#4636

Changes:
 *

### CODE REVIEW

File: cocos/core/scene-graph/base-node.ts [hunk 1]
```diff
@@ -962,7 +963,14 @@ export class BaseNode extends CCObject implements ISchedulable {
 
         const component = new constructor();
         component.node = this;
+        if (this._prefab) {
+            component.__prefab = new ComponentPrefabInfo();
+            component.__prefab.fileId = component.uuid;
+        }
```
Comment (jareguo) [inline cocos/core/scene-graph/base-node.ts]: As node's attaching process, 

we can just trigger an Editor event here: await Editor.Message.request('scene', 'link-component', nodeUuid, assetUuid);

File: cocos/core/components/component.ts [hunk 2]
```diff
@@ -175,6 +175,12 @@ class Component extends CCObject {
     @serializable
     public _enabled = true;
 
+    /**
+     * @private
+     */
+    @serializable
+    public __prefab: any = null;
```
Comment (jareguo) [inline cocos/core/components/component.ts]: Could use type ComponentPrefabInfo here

File: cocos/core/scene-graph/base-node.ts [hunk 3]
```diff
@@ -962,7 +963,14 @@ export class BaseNode extends CCObject implements ISchedulable {
 
         const component = new constructor();
         component.node = this;
+        if (this._prefab) {
```
Comment (jareguo) [inline cocos/core/scene-graph/base-node.ts]: if (EDITOR && this._prefab)

File: cocos/core/utils/prefab-helper.ts [hunk 4]
```diff
@@ -148,3 +148,16 @@ export default function syncWithPrefab (node) {
     node._localZOrder = _localZOrder;
     node._globalZOrder = _globalZOrder;
 }
+
+@ccclass('cc.ComponentPrefabInfo')
```
Comment (jareguo) [inline cocos/core/utils/prefab-helper.ts]: 类名就 cc.CompPrefabInfo 吧，短点

File: cocos/core/utils/prefab-helper.ts [hunk 5]
```diff
@@ -148,3 +148,16 @@ export default function syncWithPrefab (node) {
...
+export class ComponentPrefabInfo {
+    /**
+     * 组件在 prefabNode 中，prefabNode 从资源还原时
+     * component.__prefab.fileId 作为是否复用组件的判断依据
```
Comment (jareguo) [inline cocos/core/utils/prefab-helper.ts]: To identify current component in a prefab asset, so only needs to be unique.

File: cocos/core/scene-graph/base-node.ts (hunk 1 above)
Comment (arsen2010) [reply cocos/core/scene-graph/base-node.ts]: This mark and below marks have been adjusted.

**Review by pandamicro — APPROVED** [review]

**Pull request merged**